File: apps/mobile/modules/liquid-toasts/android/src/main/java/expo/modules/liquidtoasts/LiquidToastsModule.kt
package expo.modules.liquidtoasts

import android.app.Activity
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ProcessLifecycleOwner
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

private class InvalidArgumentsException(message: String) :
    CodedException("ERR_INVALID_ARGS", message, null)

/**
 * The JS bridge — the Android analog of `LiquidToastsModule.swift`. Decodes call
 * arguments into [ToastModel]s, drives the [ToastManager], and emits lifecycle
 * events back to JS. The Expo Modules API dispatches these handlers on the main
 * thread, so UI is touched directly.
 *
 * State ownership:
 *  - a [ToastManager] + its embedded [DeadlineScheduler] (wired via `onExpire`);
 *  - a main-confined [CoroutineScope] (SupervisorJob + Main.immediate), cancelled
 *    on destroy;
 *  - a [ProcessLifecycleOwner] observer mapping ON_STOP/ON_START to the manager's
 *    background/foreground sweep;
 *  - an [OverlayHost] installed eagerly (so the first toast gets its entrance)
 *    and rebuilt across activity changes while the manager state survives.
 *
 * Wire ack shapes match iOS exactly (see `LiquidToastsModule.swift`).
 */
class LiquidToastsModule : Module() {
    private var scope: CoroutineScope? = null
    private var manager: ToastManager? = null
    private var overlay: OverlayHost? = null
    private var overlayActivity: Activity? = null

    private val lifecycleObserver = object : DefaultLifecycleObserver {
        override fun onStart(owner: LifecycleOwner) {
            manager?.appWillEnterForeground()
        }

        override fun onStop(owner: LifecycleOwner) {
            manager?.appDidEnterBackground()
        }
    }

    override fun definition() = ModuleDefinition {
        Name("LiquidToasts")

        // Single event carrying the wire-ready payload; JS routes by `id`.
        Events("onToastEvent")

        OnCreate {
            val newScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
            scope = newScope
            val density = appContext.reactContext?.resources?.displayMetrics?.density ?: 1f
            val newManager = ToastManager(
                scope = newScope,
                decodeImage = { bytes -> ToastImageDecoder.decode(bytes, density) },
            )
            newManager.onEvent = { payload ->
                newScope.launch { sendEvent("onToastEvent", payload) }
            }
            newManager.onHaptic = { kind -> overlay?.performHaptic(kind) }
            manager = newManager

            ProcessLifecycleOwner.get().lifecycle.addObserver(lifecycleObserver)
            // Eager install so the first toast animates in.
            installOverlay()
        }

        OnDestroy {
            ProcessLifecycleOwner.get().lifecycle.removeObserver(lifecycleObserver)
            overlay?.teardown()
            overlay = null
            overlayActivity = null
            manager = null
            scope?.cancel()
            scope = null
        }

        // A JS reload gets a fresh activity attachment; rebuild the overlay
        // against it while the manager state survives.
        OnActivityEntersForeground {
            installOverlay()
        }

        /**
         * Flushes any state left over from a previous JS context. Native flushes
         * **unconditionally** on every handshake, which clears stale toasts after a
         * fast-refresh / reload (the old JS listener is gone, so those toasts must
         * be dropped silently).
         */
        AsyncFunction("handshake") { _: String ->
            requireManager().flushAll()
            installOverlay()
        }

        AsyncFunction("configure") { args: Map<String, Any?> ->
            val manager = requireManager()
            args.optInt("maxVisible")?.let { manager.maxVisible = maxOf(1, it) }
            args.optInt("maxQueue")?.let { manager.maxQueue = maxOf(1, it) }
            (args["dropPolicy"] as? String)?.let { manager.dropOldest = it != "dropNewest" }
            args.optMap("safeArea")?.let { safeArea ->
                manager.customSafeArea.value = ToastSafeArea(
                    top = maxOf(0.0, safeArea.optDouble("top") ?: 0.0).toFloat(),
                    left = maxOf(0.0, safeArea.optDouble("left") ?: 0.0).toFloat(),
                    bottom = maxOf(0.0, safeArea.optDouble("bottom") ?: 0.0).toFloat(),
                    right = maxOf(0.0, safeArea.optDouble("right") ?: 0.0).toFloat(),
                )
            }
            // defaultGlass is decoded-and-ignored (exact iOS parity).
        }

        AsyncFunction("show") { args: Map<String, Any?> ->
            val manager = requireManager()
            installOverlay()
            val id = args.optString("id")
                ?: throw InvalidArgumentsException("show: missing id/message")
            val model = ToastModel.fromWire(args, id)
                ?: throw InvalidArgumentsException("show: missing id/message")
            manager.present(model, args.byteArray("image"))
            mapOf(
                "id" to model.id,
                "accepted" to true,
                "capability" to mapOf(
                    "dynamicIslandOriginUsed" to false,
                    "glassMode" to GLASS_MODE,
                ),
            )
        }

        AsyncFunction("update") { args: Map<String, Any?> ->
            val manager = requireManager()
            val id = args.optString("id")
                ?: throw InvalidArgumentsException("update: missing id/message")
            val model = ToastModel.fromWire(args, id)
                ?: throw InvalidArgumentsException("update: missing id/message")
            val applied = manager.update(id, model, args.byteArray("image"))
            val res = mutableMapOf<String, Any?>("id" to id, "applied" to applied)
            if (!applied) res["reason"] = "unknown_id"
            res
        }

        AsyncFunction("dismiss") { id: String ->
            val ok = requireManager().dismiss(id, "manual")
            val res = mutableMapOf<String, Any?>("id" to id, "dismissed" to ok)
            if (!ok) res["reason"] = "unknown_id"
            res
        }

        AsyncFunction("dismissAll") { reason: String? ->
            mapOf("dismissedIds" to requireManager().dismissAll(reason ?: "dismissAll"))
        }

        AsyncFunction("finishAction") { id: String ->
            requireManager().finishAction(id)
        }

        /**
         * Simulates an action-button tap (drives the spinner + lifecycle). Used by
         * demos/tests that can't synthesize a real touch.
         */
        AsyncFunction("debugTriggerAction") { id: String ->
            requireManager().handleActionTap(id)
        }

        AsyncFunction("queryGeometry") {
            Geometry.snapshot(appContext.throwingActivity).toMutableMap().apply {
                this["glassMode"] = GLASS_MODE
            }
        }
    }

    /**
     * (Re)installs the overlay against the current activity. Tears down a host
     * bound to a stale activity first — the [ToastManager] survives, so
     * already-shown toasts skip their entrance via `ToastModel.hasEntered`.
     */
    private fun installOverlay() {
        val activity = appContext.activityProvider?.currentActivity ?: return
        val manager = manager ?: return
        if (overlay != null && overlayActivity === activity) {
            overlay?.bringToFront()
            return
        }
        overlay?.teardown()
        val host = OverlayHost(activity, manager)
        overlay = host
        overlayActivity = activity
        host.install()
    }

    private fun requireManager(): ToastManager =
        manager ?: throw CodedException("ERR_NOT_ATTACHED", "Module not initialized", null)

    private companion object {
        /**
         * Android renders an opaque surface (no Liquid Glass / no blur); nothing on
         * the JS side branches on this — it's an honest capability string.
         */
        const val GLASS_MODE = "opaque"
    }
}

File: apps/mobile/modules/liquid-toasts/ios/LiquidToastsModule.swift
import ExpoModulesCore
import UIKit

/// Thin bridge between JS and the native overlay. Decodes call arguments into
/// [ToastModel]s, drives [ToastManager], and emits lifecycle events back to JS.
///
/// The Expo Modules API runs async functions on a background queue unless they
/// explicitly opt into `.main`. All UI work below is therefore main-queued.
public class LiquidToastsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("LiquidToasts")

    // Single event carrying the wire-ready payload; JS routes by `id`, exactly
    // like the Flutter event channel did.
    Events("onToastEvent")

    OnCreate {
      // Install the (empty) overlay eagerly so SwiftUI has rendered the container
      // before the first `show` — otherwise the first toast appears as initial
      // content and skips its entrance transition.
      Task { @MainActor [weak self] in
        ToastOverlayHost.shared.ensureInstalled()
        self?.wireEvents()
      }
    }

    /// Flushes any state left over from a previous JS context. Native flushes
    /// **unconditionally** on every handshake, which is what clears stale toasts
    /// after a fast-refresh / reload (the old JS listener is gone, so those
    /// toasts must be dropped silently).
    AsyncFunction("handshake") { (_: String) in
      MainActor.assumeIsolated {
        let host = ToastOverlayHost.shared
        self.wireEvents()
        host.manager.flushAll()
        host.ensureInstalled()
      }
    }.runOnQueue(.main)

    AsyncFunction("configure") { (args: [String: Any]) in
      MainActor.assumeIsolated {
        let manager = ToastOverlayHost.shared.manager
        if let value = args.int("maxVisible") { manager.maxVisible = max(1, value) }
        if let value = args.int("maxQueue") { manager.maxQueue = max(1, value) }
        if let policy = args["dropPolicy"] as? String { manager.dropOldest = policy != "dropNewest" }
        if let safeArea = args["safeArea"] as? [String: Any] {
          let next = ToastSafeAreaInsets(
            top: safeArea.cgFloat("top") ?? 0,
            left: safeArea.cgFloat("left") ?? 0,
            bottom: safeArea.cgFloat("bottom") ?? 0,
            right: safeArea.cgFloat("right") ?? 0
          )
          if manager.customSafeArea != next { manager.customSafeArea = next }
        }
      }
    }.runOnQueue(.main)

    AsyncFunction("show") { (args: [String: Any]) -> [String: Any] in
      try MainActor.assumeIsolated {
        let host = ToastOverlayHost.shared
        host.ensureInstalled()
        guard let model = ToastModel(arguments: args) else {
          throw InvalidArgumentsException("show: missing id/message")
        }
        host.manager.present(model, imageData: Self.imageData(args["image"]))
        return [
          "id": model.id,
          "accepted": true,
          "capability": [
            "dynamicIslandOriginUsed": false,
            "glassMode": Capabilities.glassModeString,
          ],
        ]
      }
    }.runOnQueue(.main)

    AsyncFunction("update") { (args: [String: Any]) -> [String: Any] in
      try MainActor.assumeIsolated {
        guard let id = args["id"] as? String, let model = ToastModel(arguments: args) else {
          throw InvalidArgumentsException("update: missing id/message")
        }
        let applied = ToastOverlayHost.shared.manager.update(
          id: id, with: model, imageData: Self.imageData(args["image"]))
        var res: [String: Any] = ["id": id, "applied": applied]
        if !applied { res["reason"] = "unknown_id" }
        return res
      }
    }.runOnQueue(.main)

    AsyncFunction("dismiss") { (id: String) -> [String: Any] in
      MainActor.assumeIsolated {
        let ok = ToastOverlayHost.shared.manager.dismiss(id: id, reason: "manual")
        var res: [String: Any] = ["id": id, "dismissed": ok]
        if !ok { res["reason"] = "unknown_id" }
        return res
      }
    }.runOnQueue(.main)

    AsyncFunction("dismissAll") { (reason: String?) -> [String: Any] in
      MainActor.assumeIsolated {
        let ids = ToastOverlayHost.shared.manager.dismissAll(reason: reason ?? "dismissAll")
        return ["dismissedIds": ids]
      }
    }.runOnQueue(.main)

    AsyncFunction("finishAction") { (id: String) in
      MainActor.assumeIsolated {
        ToastOverlayHost.shared.manager.finishAction(id: id)
      }
    }.runOnQueue(.main)

    /// Simulates an action-button tap (drives the spinner + lifecycle). Used by
    /// demos/tests that can't synthesize a real touch.
    AsyncFunction("debugTriggerAction") { (id: String) in
      MainActor.assumeIsolated {
        ToastOverlayHost.shared.manager.handleAction(id: id)
      }
    }.runOnQueue(.main)

    AsyncFunction("queryGeometry") { () -> [String: Any] in
      MainActor.assumeIsolated {
        DynamicIslandGeometry.geometrySnapshot(ToastOverlayHost.activeWindow())
      }
    }.runOnQueue(.main)
  }

  /// Points the manager's event callback at this module instance's JS emitter.
  /// Re-wired on handshake so a reloaded JS context gets a live sink.
  @MainActor
  private func wireEvents() {
    ToastOverlayHost.shared.manager.onEvent = { [weak self] payload in
      self?.sendEvent("onToastEvent", payload)
    }
  }

  /// Decodes the base64 image payload. JS has no byte-array type over the
  /// bridge, so the leading image crosses as base64 and is decoded here (the
  /// actual bitmap decode still happens off-main in `ToastImageDecoder`).
  private static func imageData(_ value: Any?) -> Data? {
    guard let base64 = value as? String, !base64.isEmpty else { return nil }
    return Data(base64Encoded: base64)
  }
}

private final class InvalidArgumentsException: GenericException<String>, @unchecked Sendable {
  override var reason: String { param }
}
